lib/hammer_cli_katello/content_view_version.rb

```diff
@@ -259,6 +261,134 @@ class ExportCommand < HammerCLIKatello::SingleResourceCommand
+      failure_message _("Could not export the content view")
+
+      option "--id", "ID", _("Content View Version numeric identifier")
+      option(
+        '--export-dir',
+        'EXPORT_DIR',
+        _("Directory to put content view version export into.")
+      )
+
+      validate_options do
+        option(:option_export_dir).required
+      end
+
+      build_options
+
+      class ContentViewIdParamSource
+        def initialize(command)
+          @command = command
+        end
+
+        def get_options(_defined_options, result)
+          if result['option_id'].nil? && result['option_name']
+            result['option_id'] = @command.resolver.content_view_id(result)
+          end
+          result
+        end
+      end
+
+      def option_sources
+        super << ContentViewIdParamSource.new(self)
+      end
+
+      def execute
+        cvv = content_view_version
+        repositories = cvv['repositories'].collect do |repo|
+          repository(repo['id'])
+        end
+
+        repositories.each do |repo|
+          repo['packages'] = packages(repo['id'])
+          repo['errata'] = errata(repo['id'])
+        end
+
+        json = export_json(cvv, repositories)
+        create_tar(cvv, repositories, json)
+      end
+
+      def create_tar(cvv, repositories, json)
+        export_prefix = "export-#{cvv['id']}"
+        export_file = "#{export_prefix}.json"
+        export_repos_tar = "#{export_prefix}-repos.tar"
+        export_tar = "#{export_prefix}.tar"
+        published_repo_dir = "/var/lib/pulp/published/yum/https/repos/"
+
+        Dir.mkdir("#{options['option_export_dir']}/#{export_prefix}")
+
+        Dir.chdir(published_repo_dir) do
+          repositories.each do |repo|
+            repo_tar = "#{options['option_export_dir']}/#{export_prefix}/#{export_repos_tar}"
+            `tar cvfh #{repo_tar} #{repo['relative_path']}`
+          end
+        end
+
+        Dir.chdir("#{options['option_export_dir']}/#{export_prefix}") do
+          File.open(export_file, 'w') do |file|
+            file.write(JSON.pretty_generate(json))
+          end
+        end
+
+        Dir.chdir(options['option_export_dir']) do
+          `tar cf #{export_tar} #{export_prefix}`
+          FileUtils.rm_rf(export_prefix)
+        end
+
+        return 0
+      end
+
+      def content_view_version
+        resource_content_view_versions.call(:show, 'id' => options['option_id'])
+      end
+
+      def resource_content_view_versions
+        HammerCLIForeman.foreman_resource(:content_view_versions)
+      end
+
+      def repository(id)
+        HammerCLIForeman.foreman_resource(:repositories).call(:show, 'id' => id)
+      end
+
+      def packages(id)
+        HammerCLIForeman.foreman_resource(:packages).call(:index, 'repository_id' => id)['results']
+      end
+
+      def errata(id)
+        HammerCLIForeman.foreman_resource(:errata).call(:index, 'repository_id' => id)['results']
+      end
```

Comment:
Looks like we should define another helper. Again, I don't believe we should be constructing the json in hammer (that's why no other hammer commands require the raw data like this). I'd *much* prefer Katello be able to return the JSON structure you're after, whether via API flag in `content_view_versions#show` or a new API endpoint. But, if we really want it in hammer, let's at least put it in a helper:

i.e.

```rb
      def show(resource, options)
        HammerCLIForeman.foreman_resource(resource).call(:show, options)
      end

      def index(resource, options)
        HammerCLIForeman.foreman_resource(resource).call(:index, options)['results']
      end
```

Reply:
Sorry, I don't follow, I have methods like this already. Renaming them to index or show might make them more generic and less readable given multiple show calls are made?

Reply:
I actually think these two methods make the calls *more* readable. i.e.

```rb
# this
show(:content_view_versions, 'id' => options['option_id'])

# instead of this
HammerCLIForeman.foreman_resource(:content_view_versions).call(:show, 'id' => options['option_id'])
```

You don't think it's more readable that way? Import also does some of this type of thing, so I think having generic methods in a helper outside this class would be an improvement. In addition, you can DRY up the index logic of fetching the `['results']`, and for import, you can support ID resolution from name + other required params for show queries, which would address [this comment](https://github.com/Katello/hammer-cli-katello/pull/592/files#r222803347).

Reply:
Honestly then, I need some help how you would structure this helper. A helper per resource type? A single helper with all of these methods put into it? is it then a specific import/export helper file? or somehow made generic?

Reply:
I'm envisioning something very generic, one method per standard REST action, and one action-generic method (i.e. for [repositories#sync](https://github.com/Katello/hammer-cli-katello/pull/592/files#diff-c16d36ded485c9e83859c586343e3bb6R459) and others):

```rb
module ApipieHelper # or whatever name you think appropriate
  def show(resource, options)
    call(:show, resource, options)
  end

  def index(resource, options)
    call(:index, resource, options)['results']
  end

  def call(action, resource, options)
    HammerCLIForeman.foreman_resource(resource).call(action, options)['results']
  end
end
```

I think such a helper could also be moved over to hammer-cli-foreman with the `LocalHelper`.